### PR TITLE
update HogwildWorld with recent features, improve precision, support valid & batching

### PR DESCRIPTION
**Pull request by alexholdenmiller**

### CODE REVIEW

File: parlai/core/teachers.py [hunk 1]
```diff
@@ -112,6 +114,11 @@ class FixedDialogTeacher(Teacher):
 
 
     """
+    class AttrDict(dict):
+        def __init__(self, *args, **kwargs):
```
Comment (jaseweston) [inline parlai/core/teachers.py]: a comment explaining this?

Comment (alexholdenmiller) [reply parlai/core/teachers.py]: move to utils where it belongs and added a comment.

**Comment by alexholdenmiller** [comment]
I'm having a little bit of an issue as I'm testing the new logging code, will have to work on it some more

**Comment by alexholdenmiller** [comment]
pausing land for more testing

File: examples/train_model.py [hunk 2]
```diff
@@ -237,9 +233,12 @@ def train(self):
             # reload best validation model
             self.agent = create_agent(opt)
 
-        run_eval(self.agent, opt, 'valid', write_log=True)
-        run_eval(self.agent, opt, 'test', write_log=True)
+        _rep, wrld = run_eval(self.agent, opt, 'valid', write_log=True)
+        wrld.shutdown()  # may need to shut down threads, remote connections
```
Comment (jaseweston) [inline examples/train_model.py]: can't you do this inside run_eval func?

**Pull request merged**